test/extensions/transport_sockets/tls/ssl_socket_test.cc

```diff
@@ -4756,6 +4769,188 @@ TEST_P(SslSocketTest, UpstreamNotReadySslSocket) {
+  client_ssl_socket_factory.addReadyCb(second_callback_.AsStdFunction());
+}
+
+// Validate that secrets callbacks are invoked when secrets become ready.
+TEST_P(SslSocketTest, ServerAddSecretsReadyCallback) {
+  Stats::TestUtil::TestStore stats_store;
+  NiceMock<LocalInfo::MockLocalInfo> local_info;
+  testing::NiceMock<Server::Configuration::MockTransportSocketFactoryContext> factory_context;
+  NiceMock<Init::MockManager> init_manager;
+  NiceMock<Event::MockDispatcher> dispatcher;
+  EXPECT_CALL(factory_context, localInfo()).WillOnce(ReturnRef(local_info));
+  EXPECT_CALL(factory_context, stats()).WillOnce(ReturnRef(stats_store));
+  EXPECT_CALL(factory_context, initManager()).WillRepeatedly(ReturnRef(init_manager));
+  EXPECT_CALL(factory_context, dispatcher()).WillRepeatedly(ReturnRef(dispatcher));
+
+  envoy::extensions::transport_sockets::tls::v3::DownstreamTlsContext tls_context;
+  auto sds_secret_configs =
+      tls_context.mutable_common_tls_context()->mutable_tls_certificate_sds_secret_configs()->Add();
+  sds_secret_configs->set_name("abc.com");
+  sds_secret_configs->mutable_sds_config();
+  auto server_cfg = std::make_unique<ServerContextConfigImpl>(tls_context, factory_context);
+  EXPECT_TRUE(server_cfg->tlsCertificates().empty());
+  EXPECT_FALSE(server_cfg->isReady());
+
+  NiceMock<Ssl::MockContextManager> context_manager;
+  ServerSslSocketFactory server_ssl_socket_factory(std::move(server_cfg), context_manager,
+                                                   stats_store, std::vector<std::string>{});
+
+  // Add a secrets ready callback. It should not be invoked until onAddOrUpdateSecret() is called.
+  MockFunction<void()> mock_callback_;
+  EXPECT_CALL(mock_callback_, Call()).Times(0);
+  server_ssl_socket_factory.addReadyCb(mock_callback_.AsStdFunction());
+
+  // Call onAddOrUpdateSecret, but return a null ssl_ctx. This should not invoke the callback.
+  EXPECT_CALL(context_manager, createSslServerContext(_, _, _, _)).WillOnce(Return(nullptr));
+  server_ssl_socket_factory.onAddOrUpdateSecret();
+
+  // Now return a ssl context which should result in the callback being invoked.
+  EXPECT_CALL(mock_callback_, Call());
+  Ssl::ServerContextSharedPtr mock_context = std::make_shared<Ssl::MockServerContext>();
+  EXPECT_CALL(context_manager, createSslServerContext(_, _, _, _)).WillOnce(Return(mock_context));
+  server_ssl_socket_factory.onAddOrUpdateSecret();
+
+  // Add another callback, it should be invoked immediately.
+  MockFunction<void()> second_callback_;
+  EXPECT_CALL(second_callback_, Call());
+  server_ssl_socket_factory.addReadyCb(second_callback_.AsStdFunction());
+}
+
+// Tests adding a callback and adding secrets in parallel.  This is intended to
+// catch a race condition introduced in
+// https://github.com/envoyproxy/envoy/pull/13516 where a callback would never
+// be called due to a concurrency bug.
+TEST_P(SslSocketTest, ServerAddSecretsReadyCallbackParallel) {
```

Comment:
This and the next test are probably a bit odd since they start a thread to try to tickle a concurrency bug that has been fixed now, and only happened in ~5% of runs on my machine. This means if there is a regression for some reason we might not see it in the PR that introduces it.

Let me know if there is a better way of handling tests like this